sdk/v2/core/projects.go

```diff
@@ -0,0 +1,327 @@
+	type Alias ProjectList
+	return json.Marshal(
+		struct {
+			meta.TypeMeta `json:",inline"`
+			Alias         `json:",inline"`
+		}{
+			TypeMeta: meta.TypeMeta{
+				APIVersion: meta.APIVersion,
+				Kind:       "ProjectList",
+			},
+			Alias: (Alias)(p),
+		},
+	)
+}
+
+// Project is Brigade's fundamental configuration, management, and isolation
+// construct.
+// - Configuration: Users define Projects to pair EventSubscriptions with
+//   template WorkerSpecs.
+// - Management: Project administrators govern Project access by granting and
+//   revoking project-level Roles to/from principals (such as Users or
+//   ServiceAccounts)
+// - Isolation: All workloads (Workers and Jobs) spawned to handle a given
+//   Project's Events are isolated from other Projects' workloads in the
+//   underlying workload execution substrate.
+type Project struct {
+	// ObjectMeta contains Project metadata.
+	meta.ObjectMeta `json:"metadata"`
+	// Description is a natural language description of the Project.
+	Description string `json:"description,omitempty"`
+	// Spec is an instance of a ProjectSpec that pairs EventSubscriptions with
+	// a WorkerTemplate.
+	Spec ProjectSpec `json:"spec"`
+	// Kubernetes contains Kubernetes-specific details of the Project's
+	// environment. These details are populated by Brigade so that sufficiently
+	// authorized Kubernetes users may obtain the information needed to directly
+	// modify a Project's environment to facilitate certain advanced use cases.
+	// Clients MUST leave the value of this field nil when using the API to create
+	// or update a Project.
+	Kubernetes *KubernetesConfig `json:"kubernetes,omitempty"`
+}
+
+// MarshalJSON amends Project instances with type metadata so that clients do
+// not need to be concerned with the tedium of doing so.
+func (p Project) MarshalJSON() ([]byte, error) {
+	type Alias Project
+	return json.Marshal(
+		struct {
+			meta.TypeMeta `json:",inline"`
+			Alias         `json:",inline"`
+		}{
+			TypeMeta: meta.TypeMeta{
+				APIVersion: meta.APIVersion,
+				Kind:       "Project",
+			},
+			Alias: (Alias)(p),
+		},
+	)
+}
+
+// ProjectSpec is the technical component of a Project. It pairs
+// EventSubscriptions with a prototypical WorkerSpec that is used as a template
+// for creating new Workers.
+type ProjectSpec struct {
+	// EventSubscription defines a set of trigger conditions under which a new
+	// Worker should be created.
+	EventSubscriptions []EventSubscription `json:"eventSubscriptions,omitempty"`
+	// WorkerTemplate is a prototypical WorkerSpec.
+	WorkerTemplate WorkerSpec `json:"workerTemplate"`
+}
+
+// EventSubscription defines a set of Events of interest. ProjectSpecs utilize
+// these in defining the Events that should trigger the execution of a new
+// Worker. An Event matches a subscription if it meets ALL of the specified
+// criteria.
+type EventSubscription struct {
+	// Source specifies the origin of an Event (e.g. a gateway).
```

Comment:
I'm wondering if it is worth spelling out the behavior of each of these filters when they are left empty/uninitialized.

Reply:
They're all required. I'm fixing the docs to reflect that.

Reply:
lmk if the recent commits add sufficient clarity.

Reply:
Thanks!